console-subscriber/src/record.rs

```diff
@@ -0,0 +1,224 @@
+use serde::{
+    ser::{SerializeSeq, SerializeStruct},
+    Serialize,
+};
+use std::{
+    fs::File,
+    io,
+    path::Path,
+    sync::{Arc, Mutex},
+    time::SystemTime,
+};
+
+use console_api as proto;
+
+/// This marks the currently understood version of the recording format. This
+/// should be increased whenever the format has a breaking change that we
+/// cannot parse. Though, even better, we should probably support parsing
+/// older versions.
+///
+/// But while this is in rapid development, we can move fast and break things.
+const DATA_FORMAT_VERSION: u8 = 1;
+
+pub(crate) struct Recorder {
+    buf: Arc<Mutex<RecordBuf>>,
+
+    worker: std::thread::JoinHandle<()>,
+}
+
+struct Io {
+    buf: Arc<Mutex<RecordBuf>>,
+    file: File,
+}
+
+struct RecordBuf {
+    /// The current buffer to serialize events into.
+    bytes: Vec<u8>,
+    /// The "next" buffer that should be used when the IO thread takes the
+    /// current buffer. After flushing, the IO thread will put the buffer
+    /// back in this slot, so the allocation can be reused.
+    next: Vec<u8>,
+}
+
+#[derive(Serialize)]
+struct Header {
+    v: u8,
+}
+
+#[derive(Serialize)]
+enum Event<'a> {
+    Spawn {
+        id: u64,
+        at: SystemTime,
+        fields: SerializeFields<'a>,
+    },
+    Enter {
+        id: u64,
+        at: SystemTime,
+    },
+    Exit {
+        id: u64,
+        at: SystemTime,
+    },
+    Close {
+        id: u64,
+        at: SystemTime,
+    },
+    Waker {
+        id: u64,
+        op: super::WakeOp,
+        at: SystemTime,
+    },
+}
+
+struct SerializeFields<'a>(&'a [proto::Field]);
+
+struct SerializeField<'a>(&'a proto::Field);
+
+impl Recorder {
+    pub(crate) fn new(path: &Path) -> io::Result<Self> {
+        let buf = Arc::new(Mutex::new(RecordBuf::new()));
+        let buf2 = buf.clone();
+        let file = std::fs::File::create(path)?;
+
+        let worker = std::thread::Builder::new()
+            .name("console/subscriber/recorder/io".into())
```

Comment:
very silly nitpick, feel free to ignore: is using slashes for thread names a widespread convention? would it be "rustier" to use `::`?
```suggestion
            .name("console_subscriber::recorder::io".into())
```

Reply:
I truly have no idea if there's much of a convention... I think the only thing I ever see is separating all words with dashes, but I don't love that. The thread names is just for debugging purposes, so 🤷